### PR TITLE
[SPARK-15532] [SQL] SQLContext/HiveContext's public constructors should use SparkSession.build.getOrCreate

### PR DESCRIPTION
**Pull request by yhuai**
## What changes were proposed in this pull request?

This PR changes SQLContext/HiveContext's public constructor to use SparkSession.build.getOrCreate and removes isRootContext from SQLContext.
## How was this patch tested?

Existing tests.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -70,6 +70,27 @@ object SQLConf {
...
+      "When set to false, only one SQLContext/HiveContext is allowed to be created " +
+      "through the constructor (new SQLContexts/HiveContexts created through newSession " +
+      "method is allowed). Please note that this conf needs to be set in Spark Conf. Once " +
+      "a SQLContext/HiveContext has been created, changing the value of this conf will not " +
+      "have effect.")
+    .booleanConf
+    .createWithDefault(false)
+
+  val ALLOW_MULTIPLE_CONTEXTS_ERROR_MESSAGE =
+    SQLConfigBuilder("spark.sql.allowMultipleContexts.errorMessage")
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: This name is not great. I am open to any suggestion.

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 2]
```diff
@@ -103,7 +103,11 @@ class SparkSession private(
    * A wrapped version of this session in the form of a [[SQLContext]], for backward compatibility.
    */
   @transient
-  private[sql] val sqlContext: SQLContext = new SQLContext(this)
+  private[sql] val sqlContext: SQLContext = existingSharedState match {
+    // If we have an existing SharedState, we should not create a root SQLContext at here.
+    case Some(sharedState) => new SQLContext(this, false)
+    case None => new SQLContext(this, true)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: do we need the isRootContext flag?

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -70,6 +70,27 @@ object SQLConf {
...
 
+  val ALLOW_MULTIPLE_CONTEXTS = SQLConfigBuilder("spark.sql.allowMultipleContexts")
+    .doc("When set to true, creating multiple SQLContexts/HiveContexts is allowed. " +
+      "When set to false, only one SQLContext/HiveContext is allowed to be created " +
+      "through the constructor (new SQLContexts/HiveContexts created through newSession " +
+      "method is allowed). Please note that this conf needs to be set in Spark Conf. Once " +
+      "a SQLContext/HiveContext has been created, changing the value of this conf will not " +
+      "have effect.")
+    .booleanConf
+    .createWithDefault(false)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: this should be internal

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -70,6 +70,27 @@ object SQLConf {
...
+
+  val ALLOW_MULTIPLE_CONTEXTS_ERROR_MESSAGE =
+    SQLConfigBuilder("spark.sql.allowMultipleContexts.errorMessage")
+      .doc("The error message used by the exception when spark.sql.allowMultipleContexts is " +
+        "set to false.")
+      .stringConf
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: internal

File: sql/core/src/test/scala/org/apache/spark/sql/MultiSQLContextsSuite.scala [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+*/
+
+package org.apache.spark.sql
+
+import org.scalatest.BeforeAndAfterAll
+
+import org.apache.spark._
+import org.apache.spark.sql.internal.SQLConf
+
+class MultiSQLContextsSuite extends SparkFunSuite with BeforeAndAfterAll {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/MultiSQLContextsSuite.scala]: can we put this in SparkSessionBuilderSuite?

Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/MultiSQLContextsSuite.scala]: sure

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 4 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: done

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 2 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: I am removing isRootContext flag.

**Comment by yhuai** [comment]
oh, some ml tests failed. Let me take a look.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 3 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: done

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 6]
```diff
@@ -69,13 +67,30 @@ class SQLContext private[sql](
...
+          val errMsg = sc.conf.get(
+            SQLConf.ALLOW_MULTIPLE_CONTEXTS_ERROR_MESSAGE.key,
+            SQLConf.ALLOW_MULTIPLE_CONTEXTS_ERROR_MESSAGE.defaultValue.get)
+          throw new SparkException(errMsg)
+        } else if (!defaultSessionExists) {
+          SparkSession.setDefaultSession(session)
+        }
+      }
+      session
+    }
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: We need to do the same thing for HiveContext.

**Comment by rxin** [comment]
LGTM pending tests.

**Comment by yhuai** [comment]
Merging to master and branch 2.0.

**Pull request closed without merging**